### PR TITLE
Add .NET 4.7 to NetFxExtension

### PR DESCRIPTION
**Pull request by joaoportela**
Add .NET 4.7 installer for NetFxExtension.
Includes windows update for .NET 4.7 to work on windows 7.
Chose to use the web version of the windows update even for NetFx47Redist; This will cause the installer to fail if there is no internet connection.

I'm not sure if you guys will accept this since development moved to wix4 but our team is still using wix3, so here it is.
Based on the code from this issue: https://github.com/wixtoolset/issues/issues/5575

### CODE REVIEW

**Comment by rolix7** [comment]
I'm using the code from the commit https://github.com/wixtoolset/wix3/pull/454/commits/299d7b5e87b5ecdd15f86cceaa04f245e43a6960, but the package 'd3dcompiler64' doesn't get installed.

Here's the log:

    [09A8:09B8][2017-11-06T16:16:47]i201: Planned package: d3dcompiler64, state: Absent, default requested: Present, ba requested: None, execute: None, rollback: None, cache: No, uncache: No, dependency: None
    [09A8:09B8][2017-11-06T16:16:47]i201: Planned package: d3dcompiler86, state: Present, default requested: Present, ba requested: None, execute: None, rollback: None, cache: No, uncache: No, dependency: None
    [09A8:09B8][2017-11-06T16:16:47]i201: Planned package: NetFx47Web, state: Absent, default requested: Present, ba requested: Present, execute: Install, rollback: None, cache: Yes, uncache: No, dependency: None

My guess whould be, you can have just one prerequisite?

```xml
<WixVariable Id="WixMbaPrereqPackageId" Value="NetFx47Web" />
```

**Comment by joaoportela** [comment]
.Net 4.7 won't work without the d3dcompiler64.
Most recent windows installations already have it included.
Windows 7 may not have it, hence the windows update installation when it is missing.

Which version of windows are you using?
x86/x64? I would assume x64 from your log.

**Comment by rolix7** [comment]
I'm not completely sure, but I think you are right, it should have been x64.

**Comment by joaoportela** [comment]
Hi again! 
What's preventing this from being merged? Is it anything that I can fix?

**Comment by n1ckhansen** [comment]
I don't want to be that guy, but I'm going to be that guy.  That this PR has been sitting for 4+ months and that .NET v4.7 was released 36 days short of a year ago is a poor look.  Ongoing development priorities re v4 of WiX not withstanding, the current stable release is still v3.11.x as of today.

**Comment by ahmedflix25** [comment]
@n1ckhansen at this point i think they forget about the project

**Comment by BMurri** [comment]
When I briefly reviewed this PR 3 months ago, I had a couple of serious questions about this particular implementation, but I haven't had the time to write them out yet (one of them directly relates to rolix7's comment above).

The release model for WiX 3 has usually been about releases of Visual Studio (and .NET releases used to occur concurrently w/ VS). WiX 4 separates the various parts of WiX so they can each naturally update independently as appropriate. VS 2018 hasn't released yet, so apart from security or other equally serious bugs, a new release of v3 hasn't yet been scheduled. v4 has taken a little bit longer than expected.

**Comment by dsiswadi** [comment]
.NET 4.7.2 is already out there now. Can we also have that one too, please?

**Comment by thoemmi** [comment]
.NET Framework 4.7 was released about 2 years ago, and 4.7.1 is already 11 month old. I'd really see WiX supporting it, especially if you're using .NET Standard 2.0 libraries.

File: src/ext/NetFxExtension/wixlib/NetFx47.wxs [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        Name="redist\Windows6.1-KB4019990-x64.msu">
+        <RemotePayload CertificatePublicKey="371DD003A37769487A2A89A5A9DDB3026451B906" CertificateThumbprint="98ED99A67886D020C564923B7DF25E9AC019DF26" Description="Microsoft .NET Framework 4.7 prerequisite setup" Hash="35CC310E81EF23439BA0EC1F11D7B71DD34ADFE5" ProductName="Microsoft .NET Framework 4.7 prerequisite" Size="2873529" Version="1.0.0.0" />
+      </MsuPackage>
+
+      <!-- Update for Windows 7 x86 -->
+      <MsuPackage
+        Id="d3dcompiler86"
+        Compressed="no"
+        Permanent="yes"
+        DetectCondition="((VersionNT64) OR (VersionNT &gt; v6.1) OR (D3DCompilerExists))"
```
Comment (BMurri) [inline src/ext/NetFxExtension/wixlib/NetFx47.wxs]: DetectCondition should include only D3DCompilerExists
InstallCondition should include the rest of this line.

Also, the package for server isn't included.

File: src/ext/NetFxExtension/wixlib/NetFx47.wxs [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    <WixVariable Id="NetFx47WebInstallCondition" Value="" Overridable="yes" />
+    <WixVariable Id="NetFx47WebPackageDirectory" Value="redist\" Overridable="yes" />
+
+    <PackageGroup Id="NetFx47Web">
+      <!-- Update for Windows 7 x64 -->
+      <MsuPackage
+        Id="d3dcompiler64"
+        Compressed="no"
+        Permanent="yes"
+        DetectCondition="((NOT VersionNT64) OR (VersionNT &gt; v6.1) OR (D3DCompilerExists))"
```
Comment (BMurri) [inline src/ext/NetFxExtension/wixlib/NetFx47.wxs]: Same comments as below

File: src/ext/NetFxExtension/wixlib/NetFx47.wxs [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    -->
+
+  <?define NetFx47MinRelease = 460798 ?>
+  <?define NetFx47WebLink = http://go.microsoft.com/fwlink/?LinkId=825298 ?>
+  <?define NetFx47RedistLink = http://go.microsoft.com/fwlink/?LinkId=825302 ?>
+  <?define NetFx47EulaLink = http://referencesource.microsoft.com/license.html ?>
+
+  <Fragment>
+    <PropertyRef Id="WIXNETFX4RELEASEINSTALLED" />
+    <Property Id="WIX_IS_NETFRAMEWORK_462_OR_LATER_INSTALLED" Secure="yes" />
```
Comment (markjourdan) [inline src/ext/NetFxExtension/wixlib/NetFx47.wxs]:
```suggestion
    <Property Id="WIX_IS_NETFRAMEWORK_47_OR_LATER_INSTALLED" Secure="yes" />
```

File: src/ext/NetFxExtension/wixlib/NetFx47.wxs [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    <SetProperty Id="WIX_IS_NETFRAMEWORK_462_OR_LATER_INSTALLED" Value="1" After="AppSearch">
```
Comment (markjourdan) [inline src/ext/NetFxExtension/wixlib/NetFx47.wxs]:
```suggestion
    <SetProperty Id="WIX_IS_NETFRAMEWORK_47_OR_LATER_INSTALLED" Value="1" After="AppSearch">
```

File: src/ext/NetFxExtension/wixlib/NetFx47.wxs [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+      WIXNETFX4RELEASEINSTALLED >= "#$(var.NetFx462MinRelease)"
```
Comment (markjourdan) [inline src/ext/NetFxExtension/wixlib/NetFx47.wxs]:
```suggestion
      WIXNETFX4RELEASEINSTALLED >= "#$(var.NetFx47MinRelease)"
```

**Comment by rseanhall** [comment]
I sent a pull request for 4.7.2 and 4.8 at https://github.com/wixtoolset/wix3/pull/520. Does anyone really need 4.7 or 4.7.1? If not please close this PR. If so, we would need it accepted into the new v4 repo at https://github.com/wixtoolset/Netfx.wixext before we would consider accepting it here.

**Comment by ahmedflix25** [comment]
yes, we care

**Comment by ahmedflix25** [comment]
i see your point every app that target any version of 4.*.* should work on 4.8 just fine but there are apps that still uses them and at least they should be able to check if the right version for them or higher is installed on the machine or not the easy way

**Comment by rseanhall** [comment]
I've added the MSI properties, but not the packages, for 4.7 and 4.7.1 to my pull request. I'm closing this since we aren't going to add the packages.

**Pull request closed without merging**